File: backend/utils/currency.py
import re

_CODE_RE = re.compile(r"^[A-Z]{3}$")


def normalize_currency(code: str) -> str | None:
    if not code:
        return None
    normalized = code.strip().upper()
    if not _CODE_RE.match(normalized):
        return None
    return normalized


def parse_symbol_list(raw: str | None) -> list[str] | None:
    if raw is None or raw == "":
        return []
    parts = [part.strip() for part in raw.split(",")]
    if any(part == "" for part in parts):
        return None
    codes: list[str] = []
    for part in parts:
        normalized = normalize_currency(part)
        if not normalized:
            return None
        if normalized not in codes:
            codes.append(normalized)
    return codes

File: server/src/utils/currency.ts
const CURRENCY_CODE_REGEX = /^[A-Z]{3}$/;

export function normalizeCurrencyCode(code: string): string | null {
  const trimmed = code.trim().toUpperCase();
  if (!CURRENCY_CODE_REGEX.test(trimmed)) {
    return null;
  }
  return trimmed;
}

export function parseCurrencyList(raw: string | undefined): string[] | null {
  if (!raw) {
    return [];
  }

  const parts = raw.split(",").map((part) => part.trim());
  if (parts.length === 0 || parts.every((part) => part.length === 0)) {
    return null;
  }

  const codes: string[] = [];
  for (const part of parts) {
    if (!part) {
      return null;
    }
    const normalized = normalizeCurrencyCode(part);
    if (!normalized) {
      return null;
    }
    if (!codes.includes(normalized)) {
      codes.push(normalized);
    }
  }

  return codes;
}
